src/track/steminfoimporter.cpp

```diff
@@ -0,0 +1,144 @@
+    (uint32_t)(((uint8_t)(value)[0] << 24) | ((uint8_t)(value)[1] << 16) | \
+            ((uint8_t)(value)[2] << 8) | (uint8_t)(value)[3])
+
+namespace mixxx {
+
+namespace {
+
+const mixxx::Logger kLogger("StemInfoImporter");
+constexpr int kSupportedStemVersion = 1;
+const QStringList kStemFileExtensions = {".stem.mp4", ".stem.m4a"};
+
+const uint32_t kAtomHeaderSize = 8; // 4 bytes (unsigned integer) + 4 char
+const uint32_t kStemManifestAtomPath[] = {
+        ATOM_TYPE("moov"),
+        ATOM_TYPE("udta"),
+        ATOM_TYPE("stem"),
+        0 // This indicate end of the path
+};
+
+/// @brief Seek the reader the atom requested.
+/// @param reader The IODevice to search MP4 atom in
+/// @param path The list of atom to traverse in the tree, from top to bottom.
+/// Must be null terminated
+/// @param boxSize The size of the box currently under the cursor to focus the
+/// search in
+/// @param pathIdx The level of the tree currently search (index of path)
+/// @return the size of the box data if found, -1 otherwise
+uint32_t seekTillAtom(QIODevice& reader,
+        const uint32_t path[],
+        uint32_t boxSize = 0,
+        uint32_t pathIdx = 0) {
+    if (!path[pathIdx]) {
+        return boxSize;
+    }
+
+    uint32_t byteRead = 0;
+    char buffer[kAtomHeaderSize];
+    while (!reader.atEnd() || (boxSize && byteRead >= boxSize)) {
+        reader.read(buffer, kAtomHeaderSize);
+        byteRead += boxSize;
+        if (ATOM_TYPE(buffer + 4) == path[pathIdx]) {
+            return seekTillAtom(reader, path, ATOM_TYPE(buffer) - kAtomHeaderSize, pathIdx + 1);
+        }
+        reader.skip(ATOM_TYPE(buffer) - kAtomHeaderSize);
+    }
+    return -1;
+}
+
+} // anonymous namespace
+
+// static
+bool StemInfoImporter::isStemFile(
+        const QString& fileName) {
+    for (const QString& ext : kStemFileExtensions) {
+        if (fileName.endsWith(ext)) {
+            return true;
+        }
+    }
+    return false;
+}
+
+QList<StemInfo> StemInfoImporter::importStemInfos(
+        const QString& filePath) {
+    // Fetch the STEM manifest which contain stream details
+    auto file = QFile(filePath);
+    if (!file.open(QIODeviceBase::ReadOnly)) {
+        kLogger.warning()
+                << "Failed to open input file"
+                << filePath;
+        return {};
+    }
+
+    uint32_t manifestSize;
+    if (!(manifestSize = seekTillAtom(file, kStemManifestAtomPath))) {
+        kLogger.warning()
+                << "Failed to find the steam manifest in the file"
+                << filePath;
+        return {};
+    }
+
+    auto jsonData = QJsonDocument::fromJson(file.read(manifestSize));
+    VERIFY_OR_DEBUG_ASSERT(jsonData.isObject()) {
+        kLogger.warning()
+                << "Failed to extract the manifest data"
+                << filePath;
+        return {};
+    };
+
+    auto json = jsonData.object();
+
+    // Check the manifest version
+    auto version = json.value("version").toInteger(-1);
```

Comment:
Can you add here a typical JSON as a source code comment as a reference? 
Maybe also a link to the specification.

Reply:
Maybe we could add the JSON from src/test/stems/test.stem.mp4 as .json file next to it?

Reply:
Done